src/common/ast-path.js

```diff
@@ -1,35 +1,75 @@
+import assert from "node:assert";
 import getLast from "../utils/get-last.js";
-
-function getNodeHelper(path, count) {
-  const stackIndex = getNodeStackIndexHelper(path.stack, count);
-  return stackIndex === -1 ? null : path.stack[stackIndex];
-}
-
-function getNodeStackIndexHelper(stack, count) {
-  for (let i = stack.length - 1; i >= 0; i -= 2) {
-    const value = stack[i];
-    if (value && !Array.isArray(value) && --count < 0) {
-      return i;
-    }
-  }
-  return -1;
-}
+import { getPenultimate } from "./util.js";
 
 class AstPath {
   constructor(value) {
     this.stack = [value];
   }
 
+  get isInArray() {
+    const { stack } = this;
+    return Array.isArray(stack[stack.length - 3]);
+  }
+
+  get key() {
+    const { stack, isInArray } = this;
+    return stack[stack.length - (isInArray ? 4 : 2)] ?? null;
+  }
+
+  get index() {
+    assert(this.isInArray);
+    return this.isInArray ? getPenultimate(this.stack) : null;
```

Comment:
Let's just let it return `null` if not in an array. This might make destructuring simpler: `const { key, index, node } = path;`.

Reply:
I was hoping check `.isInArray` first if not sure.

Reply:
But your point make sense.

Reply:
That'd make code more verbose. E.g. we want be able to do this:
```js
function hasPrettierIgnore(path) {
  return path.index > 0 && isPrettierIgnore(path.previous) === "next";
}
```